backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -280,5 +264,104 @@ bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
     return false;
 }
 
+const IR::Node* AddNewMetadataFields::preorder(IR::DpdkStructType *st) {
+    if (isMetadataStruct(st)) {
+        for (auto nf : newMetadataFields) {
+                st->fields.push_back(nf);
+        }
+        auto newSt = new IR::DpdkStructType(st->srcInfo, st->name,
+                                          st->annotations, st->fields);
+        newMetadataFields.clear();
+        return newSt;
+    }
+    return st;
+}
+
+const IR::Node* DirectionToRegRead::preorder(IR::DpdkAsmProgram *p) {
+        p->externDeclarations.push_back(addRegDeclInstance(registerInstanceName));
+        return p;
+     }
+
+IR::DpdkExternDeclaration* DirectionToRegRead::addRegDeclInstance(cstring instanceName) {
+    auto typepath = new IR::Path("Register");
+    auto type = new IR::Type_Name(typepath);
+    auto typeargs = new IR::Vector<IR::Type>({IR::Type::Bits::get(32),
+                                              IR::Type::Bits::get(32)});
+    auto spectype = new IR::Type_Specialized(type, typeargs);
+    auto args = new IR::Vector<IR::Argument>();
+    args->push_back(new IR::Argument(new IR::Constant(IR::Type::Bits::get(32), 1)));
+    auto annot = IR::Annotations::empty;
+    annot->addAnnotationIfNew(IR::Annotation::nameAnnotation,
+                              new IR::StringLiteral(instanceName));
+    auto decl = new IR::DpdkExternDeclaration(instanceName, annot, spectype, args,
+                    nullptr);
+    return decl;
+}
+
+void DirectionToRegRead::addMetadataField(cstring fieldName) {
+    if (newFieldName.count(fieldName))
+        return;
+     newMetadataFields.push_back(new IR::StructField(IR::ID(fieldName),
+                                 IR::Type::Bits::get(32)));
+     newFieldName.insert(fieldName);
+}
+
+bool DirectionToRegRead::isDirection(const IR::Member *m) {
+    if (m == nullptr)
+        return false;
+     return m->member.name == "pna_main_input_metadata_direction"
+         || m->member.name == "pna_pre_input_metadata_direction"
+         || m->member.name == "pna_main_parser_input_metadata_direction";
+}
+
+const IR::Node *DirectionToRegRead::postorder(IR::DpdkListStatement *l) {
+    l->statements = replaceDirectionWithRegRead(l->statements);
+    newStmts.clear();
+    return l;
+}
+
+void DirectionToRegRead::replaceDirection(const IR::Member *m) {
```

Comment:
You can add a function level comment here explaining the intended transformation (Similar to what I had added in the test file)